### PR TITLE
Split methods and streams in service info

### LINKED ISSUE #689: Add function to enumerate all methods
While working with Prometheus gRPC interceptors, I encountered a need to access all full service names registered with the given server. (see https://github.com/mwitkow/go-grpc-prometheus/issues/5#issuecomment-219898272) 

This would avoid some pitfalls with Prometheus data models, and make counter increments faster (as it would allow for pre-computing a lookup table).

@iamqizhao would something like this be upstream-able? :)

### PR DESCRIPTION
**Pull request by menghanl**
fixes #689

### CODE REVIEW

**Comment by menghanl** [comment]
@mwitkow FYI

**Comment by mwitkow** [comment]
Thanks @menghanl 
Would it be possible to surface the `IsServerStream` or `IsClientStream` similar to `grpc.StreamServerInfo`? Prometheus interceptor relies on these to decide how to label the RPC
https://github.com/mwitkow/go-grpc-prometheus/blob/master/server_reporter.go#L16

**Comment by menghanl** [comment]
@mwitkow Added `StreamInfo` for streaming RPCs, PTAL.

**Comment by mwitkow** [comment]
Awesome, thank you! :)

File: server.go [hunk 1]
```diff
@@ -245,10 +245,22 @@ func (s *Server) register(sd *ServiceDesc, ss interface{}) {
 	s.m[sd.ServiceName] = srv
 }
 
-// ServiceInfo contains method names and metadata for a service.
+// StreamInfo contains information about a streaming RPC.
+type StreamInfo struct {
+	// Name is the RPC name only, without the service name or package name.
```
Comment (iamqizhao) [inline server.go]: MethodName?

File: server.go [hunk 2]
```diff
@@ -245,10 +245,22 @@ func (s *Server) register(sd *ServiceDesc, ss interface{}) {
...
+	IsServerStream bool
+}
+
+// ServiceInfo contains unary RPC names, streaming RPC infos and metadata for a service.
 type ServiceInfo struct {
-	// Methods are method names only, without the service name or package name.
+	// Methods are unary RPC names only, without the service name or package name.
 	Methods []string
+	// Streams are streaming RPC names and streaming types.
+	Streams []*StreamInfo
```
Comment (iamqizhao) [inline server.go]: merge the info for unary and stremaing like

```
Unary []Info
Stream []Info
```
For unary rpc, both IsClientStream and IsServerStream are false.

File: reflection/serverreflection.go [hunk 3]
```diff
@@ -214,19 +214,19 @@ func (s *serverReflectionServer) serviceMetadataForSymbol(name string) (interfac
 		return nil, fmt.Errorf("unknown symbol: %v", name)
 	}
 
-	// Search for method in info.
+	// Search for method name in info.Methods.
```
Comment (iamqizhao) [inline reflection/serverreflection.go]: Search the method name in info.Methods.

File: server.go [hunk 4]
```diff
@@ -245,10 +245,19 @@ func (s *Server) register(sd *ServiceDesc, ss interface{}) {
 	s.m[sd.ServiceName] = srv
 }
 
-// ServiceInfo contains method names and metadata for a service.
+// MethodInfo contains information about an RPC.
```
Comment (iamqizhao) [inline server.go]: MethodInfo contains the information of an RPC including its method name and type.

**Pull request merged**